### ISSUE TITLE
resize_tensor_input does not change shape in tflite interpreter for efficientdet models

### ISSUE THREAD

**Issue by ivanlen** [tensorflow/tensorflow#50957]
**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): NO
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Linux Ubuntu 20.04 LTS
- TensorFlow installed from (source or binary): conda tf
- TensorFlow version (use command below): 2.4.1
- Python version: 3.7.8

**Describe the current behavior**
I am trying to modify the batch size of a tflite efficientdet model following this [official tf guide](https://colab.research.google.com/github/frogermcs/TFLite-Tester/blob/master/notebooks/Testing_TFLite_model.ipynb#scrollTo=VXX8lJkPRYFq)
However I cannot modify the output shapes.

I download the model from the [TF hub](https://tfhub.dev/tensorflow/lite-model/efficientdet/lite2/detection/default/1) and load them into `model_path`.

```python
tflite_interpreter = tf.lite.Interpreter(model_path, num_threads=1)
input_details = tflite_interpreter.get_input_details()
output_details = tflite_interpreter.get_output_details()
```

Then just to be sure that everything works fine, I set `batch_size=1` which is the default:
```python
batch_size = 1
# tflite_interpreter.resize_tensor_input(output_details[0]['index'], (batch_size, 25, 4,))
# tflite_interpreter.resize_tensor_input(output_details[1]['index'], (batch_size, 25,))
# tflite_interpreter.resize_tensor_input(output_details[2]['index'], (batch_size, 25,))
# tflite_interpreter.resize_tensor_input(output_details[3]['index'], (batch_size,))
tflite_interpreter.resize_tensor_input(input_details[0]['index'], (batch_size, model_res, model_res, 3))
tflite_interpreter.allocate_tensors()

# print(input_details)
for od in tflite_interpreter.get_input_details():
    print("\n".join("{}\t{}".format(k, v) for k, v in od.items()))
    print('___')
# output details
for od in tflite_interpreter.get_output_details():
    print("\n".join("{}\t{}".format(k, v) for k, v in od.items()))
    print('___')
```
I obtain the following output
```
name	serving_default_images:0
index	0
shape	[  1 448 448   3]
shape_signature	[  1 448 448   3]
dtype	<class 'numpy.uint8'>
quantization	(0.0078125, 127)
quantization_parameters	{'scales': array([0.0078125], dtype=float32), 'zero_points': array([127], dtype=int32), 'quantized_dimension': 0}
sparsity_parameters	{}
___
name	StatefulPartitionedCall:3
index	782
shape	[ 1 25  4]
shape_signature	[ 1 25  4]
dtype	<class 'numpy.float32'>
quantization	(0.0, 0)
quantization_parameters	{'scales': array([], dtype=float32), 'zero_points': array([], dtype=int32), 'quantized_dimension': 0}
sparsity_parameters	{}
___
name	StatefulPartitionedCall:2
index	783
shape	[ 1 25]
shape_signature	[ 1 25]
dtype	<class 'numpy.float32'>
quantization	(0.0, 0)
quantization_parameters	{'scales': array([], dtype=float32), 'zero_points': array([], dtype=int32), 'quantized_dimension': 0}
sparsity_parameters	{}
___
name	StatefulPartitionedCall:1
index	784
shape	[ 1 25]
shape_signature	[ 1 25]
dtype	<class 'numpy.float32'>
quantization	(0.0, 0)
quantization_parameters	{'scales': array([], dtype=float32), 'zero_points': array([], dtype=int32), 'quantized_dimension': 0}
sparsity_parameters	{}
___
name	StatefulPartitionedCall:0
index	785
shape	[1]
shape_signature	[1]
dtype	<class 'numpy.float32'>
quantization	(0.0, 0)
quantization_parameters	{'scales': array([], dtype=float32), 'zero_points': array([], dtype=int32), 'quantized_dimension': 0}
sparsity_parameters	{}
___
```
so far so good... I kwow that tflite efficient det produces a max of 25 outputs.

Now I try to change the batch size. The input shape changes (now batch size is 4) but nothing changes for the output details.

```python
batch_size = 4

tflite_interpreter = tf.lite.Interpreter(model_path, num_threads=1)
tflite_interpreter.resize_tensor_input(input_details[0]['index'], (batch_size, model_res, model_res, 3))
tflite_interpreter.resize_tensor_input(output_details[0]['index'], (batch_size, 25, 4,))
tflite_interpreter.resize_tensor_input(output_details[1]['index'], (batch_size, 25,))
tflite_interpreter.resize_tensor_input(output_details[2]['index'], (batch_size, 25,))
tflite_interpreter.resize_tensor_input(output_details[3]['index'], (batch_size,))
tflite_interpreter.allocate_tensors()

# print(input_details)
for od in tflite_interpreter.get_input_details():
    print("\n".join("{}\t{}".format(k, v) for k, v in od.items()))
    print('___')
# output details
for od in tflite_interpreter.get_output_details():
    print("\n".join("{}\t{}".format(k, v) for k, v in od.items()))
    print('___')
```
```
name	serving_default_images:0
index	0
shape	[  4 448 448   3]
shape_signature	[  4 448 448   3]
dtype	<class 'numpy.uint8'>
quantization	(0.0078125, 127)
quantization_parameters	{'scales': array([0.0078125], dtype=float32), 'zero_points': array([127], dtype=int32), 'quantized_dimension': 0}
sparsity_parameters	{}
___
name	StatefulPartitionedCall:3
index	782
shape	[ 1 25  4]
shape_signature	[ 1 25  4]
dtype	<class 'numpy.float32'>
quantization	(0.0, 0)
quantization_parameters	{'scales': array([], dtype=float32), 'zero_points': array([], dtype=int32), 'quantized_dimension': 0}
sparsity_parameters	{}
___
name	StatefulPartitionedCall:2
index	783
shape	[ 1 25]
shape_signature	[ 1 25]
dtype	<class 'numpy.float32'>
quantization	(0.0, 0)
quantization_parameters	{'scales': array([], dtype=float32), 'zero_points': array([], dtype=int32), 'quantized_dimension': 0}
sparsity_parameters	{}
___
name	StatefulPartitionedCall:1
index	784
shape	[ 1 25]
shape_signature	[ 1 25]
dtype	<class 'numpy.float32'>
quantization	(0.0, 0)
quantization_parameters	{'scales': array([], dtype=float32), 'zero_points': array([], dtype=int32), 'quantized_dimension': 0}
sparsity_parameters	{}
___
name	StatefulPartitionedCall:0
index	785
shape	[1]
shape_signature	[1]
dtype	<class 'numpy.float32'>
quantization	(0.0, 0)
quantization_parameters	{'scales': array([], dtype=float32), 'zero_points': array([], dtype=int32), 'quantized_dimension': 0}
sparsity_parameters	{}
___
```

I tried this approach with different versions of the efficientdet model (0, 1, 2, 3, 4) changing their respectively  input resolutions accordingly, and still have the same problem.

**Describe the expected behavior**
I expect that the output shape is 4 as I set.

**Comment by srjoglekar246** [comment]
TFLite uses a custom implementation of Non Max Suppression, which is the postprocessing step in EfficientDet. Currently, it does not support batch size > 1.

We should probably take a few steps to address this:
1. Return an error during allocate_tensors() (from the kernel's Prepare method) if the input tensors have batch size > 1. (Right now it will fail at inference time, but there is no Prepare-time check).
2. Possibly add support to the kernel for batch size > 1.

@ziyeqinghan Do you know if the rest of the EfficientDet graph supports batch size > 1? In that case the only change we need is on the NMS kernel.

@ivanlen Unfortunately the only solution until we implement this, is to run per-image inference instead of batched. (On device, there may not be a huge difference between the two - but thats just a hunch...feel free to run experiments and report back if per-image inference is very slow)

**Comment by ivanlen** [comment]
@srjoglekar246 thanks for the clarification.

Actually, per-image inference is very slow. That is why I was trying to speed up the process by using a batch size > 1.
Do you have any suggestion on how to speed up the inference times?

Cheers,

Cheers.

**Comment by srjoglekar246** [comment]
Are you trying this on a desktop only, or do you want to deploy to mobile phones? THe solution varies based on device & API you use :-)

**Comment by ivanlen** [comment]
Both actually.
First I want to try different efficiendet models: make some inference with a test set and get some metrics.
Then, I would like to deploy the best model into mobile phones. On mobile I am going to run with a batch size=1, but on desktop I would like to test as fast as I can.

Currently inferencing with batch size = 1 takes like 4/5 seconds per image which is really slow.

**Comment by srjoglekar246** [comment]
Yeah, TFLite kernels aren't optimized for desktop inference, so they can be slower than the corresponding TF versions. A few ideas for desktop:

1. Try running the Interpreter with num_threads = 2 or 4. That *might* improve latency depending on your device, but its not guaranteed.
2. Since the EfficientDet model itself is small, you can just create multiple Interpreters & run them parallely in threads: TFLite RAM overhead typically isn't that high since we reuse tensor memory whenever possible.

On device, you can use a [delegate](https://www.tensorflow.org/lite/performance/delegates) that will significantly reduce your inference time. You can try our [benchmark tool](https://www.tensorflow.org/lite/performance/measurement) with [delegate-specific parameters](https://www.tensorflow.org/lite/performance/delegates#task-agnostic_evaluation) to figure out what works best.

**Comment by ivanlen** [comment]
Thanks for your feedback.

1. I tried to run with num_threads>1 but to be honest I didn't find an overall improvement, still the inference is quite slow for a test set of 2k images. 
2. I will try this option.

Since num_threads > 1 doesn't scale very good (at least in my case) I think that option 2 probably is the most suitable for my case.

I will investigate on the delegate, thanks.
Regarding the benchmark tool, I already used it (without trying delegate specific parameters, just with the default config). It's funny that on mobile devices the inference time in one thread is about 1 seconds per image while in my computer i8 CPU it takes about 4/5 seconds per images. But as you said, I guess that this has to be that they are optimized for mobile devices.

Thanks again for the feedback.
